pkg/controller/migmigration/stage.go

```diff
@@ -0,0 +1,415 @@
+package migmigration
+
+import (
+	"context"
+	"reflect"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	migref "github.com/konveyor/mig-controller/pkg/reference"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// StagePod - wrapper for stage pod, allowing to compare  two stage pods for equality
+type StagePod struct {
+	corev1.Pod
+}
+
+// StagePodList - a list of stage pods, with built-in stage pod deduplication
+type StagePodList []StagePod
+
+// BuildStagePods - creates a list of stage pods from a list of pods
+func BuildStagePods(list *corev1.PodList) StagePodList {
+	stagePods := StagePodList{}
+	for _, pod := range list.Items {
+		stagePods.merge(buildStagePodFromPod(migref.ObjectKey(&pod), pod.Labels, &pod))
+	}
+	return stagePods
+}
+
+func (p StagePod) equals(pod StagePod) bool {
+	for _, volume := range p.Spec.Volumes {
+		found := false
+		for _, targetVolume := range pod.Spec.Volumes {
+			if reflect.DeepEqual(volume.VolumeSource, targetVolume.VolumeSource) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+func (l *StagePodList) contains(pod StagePod) bool {
+	for _, srcPod := range *l {
+		if pod.equals(srcPod) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (l *StagePodList) merge(list ...StagePod) {
+	for _, pod := range list {
+		if !l.contains(pod) {
+			*l = append(*l, pod)
+		}
+	}
+}
+
+func (l *StagePodList) create(client k8sclient.Client) (int, error) {
+	existingPods := StagePodList{}
+	if len(*l) > 0 {
+		existingPods.list(client, (*l)[0].Labels)
+	}
+	counter := 0
+	for _, stagePod := range *l {
+		if existingPods.contains(stagePod) {
+			continue
+		}
+		err := client.Create(context.TODO(), &stagePod.Pod)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			return 0, err
+		}
+		counter++
+	}
+
+	return counter + len(existingPods), nil
+}
+
+func (l *StagePodList) list(client k8sclient.Client, labels map[string]string) error {
+	podList := corev1.PodList{}
+	options := k8sclient.MatchingLabels(labels)
+	err := client.List(context.TODO(), options, &podList)
+	if err != nil {
+		return err
+	}
+	for _, pod := range podList.Items {
+		l.merge(buildStagePodFromPod(migref.ObjectKey(&pod), pod.Labels, &pod))
+	}
+
+	return nil
+}
+
+func (t *Task) ensureStagePodsFromOrphanedPVCs() error {
+	existingStagePods := StagePodList{}
+	stagePods := StagePodList{}
+	client, err := t.getSourceClient()
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	err = existingStagePods.list(client, t.stagePodLabels())
```

Comment:
So it looks like we're comparing against the stage pods from running pods but not from templates here. Shouldn't we compare against both before creating a pod for a disconnected PVC?

Reply:
`.list` is looking only for created stage pods - https://github.com/konveyor/mig-controller/pull/485/files#diff-656fd52107ab09d74773450b8b7d546bR90

Each time `stagePodList.merge` method is called, it is doing de-duplication with existing list within `merge` method.  And the stage pods from templates were already created in the cluster during previous phase, so we have all of them stored in a single location at that moment.